src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonIListConverter.cs

```diff
@@ -0,0 +1,102 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections;
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// Converter for <cref>System.Collections.IList</cref>.
+    internal sealed class JsonIListConverter<TCollection> : JsonIEnumerableDefaultConverter<TCollection, object>
+        where TCollection : IList
+    {
+        protected override void Add(object value, ref ReadStack state)
+        {
+            Debug.Assert(state.Current.ReturnValue is IList);
+            ((IList)state.Current.ReturnValue!).Add(value);
+        }
+
+        protected override void CreateCollection(ref ReadStack state, JsonSerializerOptions options)
+        {
+            JsonClassInfo classInfo = state.Current.JsonClassInfo;
+
+            if (TypeToConvert.IsInterface || TypeToConvert.IsAbstract)
+            {
+                if (!TypeToConvert.IsAssignableFrom(RuntimeType))
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+
+                state.Current.ReturnValue = new List<object>();
+            }
+            else
+            {
+                if (classInfo.CreateObject == null)
+                {
+                    ThrowHelper.ThrowNotSupportedException_DeserializeNoParameterlessConstructor(TypeToConvert);
+                }
+
+                TCollection returnValue = (TCollection)classInfo.CreateObject!()!;
+
+                if (returnValue.IsReadOnly)
+                {
+                    ThrowHelper.ThrowNotSupportedException_SerializationNotSupported(TypeToConvert);
+                }
+
+                state.Current.ReturnValue = returnValue;
+            }
+        }
+
+        protected override bool OnWriteResume(Utf8JsonWriter writer, TCollection value, JsonSerializerOptions options, ref WriteStack state)
+        {
+            IEnumerator enumerator;
+            if (state.Current.CollectionEnumerator == null)
+            {
+                enumerator = value.GetEnumerator();
+                if (!enumerator.MoveNext())
+                {
+                    return true;
+                }
+            }
+            else
+            {
+                enumerator = state.Current.CollectionEnumerator;
+            }
+
+            JsonConverter<object> converter = GetElementConverter(ref state);
+            do
+            {
+                if (ShouldFlush(writer, ref state))
+                {
+                    state.Current.CollectionEnumerator = enumerator;
+                    return false;
+                }
+
+                object? element = enumerator.Current;
+
+                if (!converter.TryWrite(writer, element!, options, ref state))
```

Comment:
Same here/elsewhere. We should audit and avoid null-forgiving (`!`) on all calls to `.TryWrite`. Can the converter be of `JsonConverter<object?>`

Reply:
Same with `.TryRead`

Reply:
Why was this resolved? Still need to remove the `!`.

Reply:
This is calling a generic method which can't specify nullability for the `value` parameter since `T` can either be a struct or object.